File: binaryTree/1339_maximum_product_of_splitted_binary_tree.py
# O(n) time | O(n) space 
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        MOD = 10 ** 9 + 7
        self.ans = totalSum = 0

        def total(node):
            if not node: return 0
            return node.val + helper(node.left) + helper(node.right)
        
        totalSum = total(root)

        def traverse(node):
            if not node: return 0
            subSum = node.val + traverse(node.left) + traverse(node.right)
            productSum = (totalSum - subSum) * subSum
            self.ans = max(self.ans, productSum)
            return subtreeSum
            
        traverse(root)

        return self.ans % MOD

File: binaryTree/1339_maximum_product_of_splitted_binary_tree.go
// O(n) time | O(n) space 
func maxProduct(root *TreeNode) int {
    MOD := math.Pow(10, 9) + 7
    totalSum := sum(root)
    ans := 0
    traverse(root, totalSum, &ans)
    return ans % int(MOD)
}

func sum(node *TreeNode) int {
    if node == nil {
        return 0
    }
    return node.Val + sum(node.Left) + sum(node.Right)
}

func traverse(node *TreeNode, totalSum int, ans *int) int {
    if node == nil {
        return 0
    }
    subSum := node.Val + traverse(node.Left, totalSum, ans) + traverse(node.Right, totalSum, ans)
    productSum := (totalSum - subSum) * subSum
    if productSum > *ans {
        *ans = productSum
    }
    return subSum
}
